NIFIREG-395 - Implemented the ability to import and export versioned flows through the UI

mtien-apache opened the pull request:
Thank you for submitting a contribution to Apache NiFi Registry.

Please provide a short description of the PR here:

#### Description of PR

- [WIP] This PR adds the ability for a user to take a previously saved versionedFlowSnapshot JSON file and import it as a new flow definition to an existing bucket or import it as the next new version of an existing flow through the UI. 

- This PR also adds the ability to export a versionedFlowSnapshot as a JSON file, which removes snapshot metadata associated with the given Registry. 

- New dialogs have been created to handle import and export.

- The REST endpoints added in `BucketFlowResource` are - 
     - `importFlow` - `/buckets/{bucketId}/flows/import`
     - `importVersionedFlow` - `/buckets/{bucketId}/flows/{flowId}/versions/import`
     - `exportVersionedFlow`- `/buckets/{bucketId}/flows/{flowId}/versions/{versionNumber: \d+}/export`

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFIREG-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [ ] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi-registry` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-registry-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-registry-assembly`?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

mtien-apache commented:
This PR is still WIP, but is open for initial review. I still need to add integration tests and resolve some styling issues, specifically with text alignment in the dialogs. 

@exceptionfactory @thenatog @bbende @pvillard31 Would you please review? 

@scottyaslan @sardell @akocsis @andras Would you please review the frontend code?

Thank you!

bbende commented on nifi-registry-core/nifi-registry-web-api/src/main/java/org/apache/nifi/registry/web/api/BucketFlowResource.java:
```diff
@@ -291,6 +303,129 @@ public Response createFlowVersion(
+        versionedFlowSnapshot = deserializeVersionedFlowSnapshot(in);
+
+        // clear or set the necessary snapShotMetadata
+        if (versionedFlowSnapshot.getSnapshotMetadata() != null) {
```
I think we always want to clear these metadata values for the case where someone has json from another registry that was obtained before your new export existed. We probably also need to protect against the metadata being null because I think it is null coming from "Download Flow" in NiFi, so maybe always set to a new metadata and then apply anything on top of that?

mtien-apache replied:
Good idea to set new metadata first - I revised this to create a new snapshotMetadata then set any necessary properties from there.

bbende commented on nifi-registry-core/nifi-registry-web-api/src/main/java/org/apache/nifi/registry/web/api/BucketFlowResource.java:
```diff
@@ -385,6 +520,64 @@ public Response getLatestFlowVersionMetadata(
+            throw new IllegalArgumentException("The version number is required.");
+        }
+
+        final VersionedFlowSnapshot versionedFlowSnapshot = serviceFacade.getFlowSnapshot(bucketId, flowId, versionNumber);
```
This is kind of minor, but it might be nice to move most of this logic behind the service facade to a new method like exportFlowSnapshot(bucketId, flowId, versionNumber). We tried to make it so most of the logic happens behind the service facade, and the REST end-points don't do too much.

There is also an example of an existing file-download for extension bundles, which uses the StreamingContent/StreamingOutput, it is not as big a deal for this case since the JSON probably won't ever be too large, but in case you want to try and use this approach...

https://github.com/apache/nifi-registry/blob/main/nifi-registry-core/nifi-registry-web-api/src/main/java/org/apache/nifi/registry/web/api/ExtensionRepoResource.java#L437-L445

mtien-apache replied:
I refactored the some of this logic into the service facade.

bbende reviewed (commented):
@mtien-apache nice work, this is going to be very helpful!

I left a few inline comments, but here are some overall results from testing with a secured registry where I authenticated with LDAP (i.e. using JWTs)...

Using JSON obtained from NiFi "Download flow definition"
- Initial import worked correctly
- Trying to import a second version of the flow resulted in an error "Version must be greater than zero, or use -1 to indicate latest version"
- Trying to download a version of the flow popped up the file download dialog with the name "versionedFlow", after clicking save Chrome showed a download for "versionedFlow.txt" and said "Failed - Needs authorization"

Using JSON obtained from another NiFi Registry (i.e. it had the snapshot metadata populated from the other registry):
- Import New Flow produced an error "No applicable policies could be found. Contact the system administrator.", I think because it was trying to store it based on the previous metadata that was never cleared out

exceptionfactory reviewed (commented):
Thanks for the work on this feature @mtien-apache!  Noted a few questions, will run through another review when it is no longer in draft status.

mtien-apache commented on nifi-registry-core/nifi-registry-web-api/pom.xml:
```diff
@@ -423,6 +423,16 @@
+        </dependency>
+        <dependency>
+            <groupId>com.google.protobuf</groupId>
+            <artifactId>protobuf-java</artifactId>
```
Removed - this was accidentally leftover from some previous work.

mtien-apache commented on nifi-registry-core/nifi-registry-web-api/pom.xml:
```diff
@@ -423,6 +423,16 @@
+            <artifactId>protobuf-java</artifactId>
+            <version>3.11.4</version>
+            <scope>compile</scope>
+        </dependency>
```
Removed - see above comment.

mtien-apache commented on nifi-registry-core/nifi-registry-web-api/src/main/java/org/apache/nifi/registry/web/api/BucketFlowResource.java:
```diff
@@ -385,6 +520,64 @@ public Response getLatestFlowVersionMetadata(
     }
 
+    @GET
+    @Path("{flowId}/versions/{versionNumber: \\d+}/export")
```
I don't think the space is necessary, but decided to follow the format of the other endpoints.

mtien-apache commented on nifi-registry-core/nifi-registry-web-api/src/main/java/org/apache/nifi/registry/web/api/BucketFlowResource.java:
```diff
@@ -385,6 +520,64 @@ public Response getLatestFlowVersionMetadata(
+
+        final String versionedFlowSnapshotJsonString = serializeToJson(versionedFlowSnapshot);
+
+        return generateOkResponse(versionedFlowSnapshotJsonString).header(HttpHeaders.CONTENT_DISPOSITION, String.format("attachment; filename=\"%s\"", filename)).build();
```
Agreed that's easier to read. I've made the suggested changes. Thanks!

mtien-apache commented on nifi-registry-core/nifi-registry-web-api/src/main/java/org/apache/nifi/registry/web/api/BucketFlowResource.java:
```diff
@@ -291,6 +303,129 @@ public Response createFlowVersion(
+        final VersionedFlowSnapshot createdSnapshot = serviceFacade.createFlowSnapshot(versionedFlowSnapshot);
+        publish(EventFactory.flowVersionCreated(createdSnapshot));
+
+        return Response.status(Response.Status.OK).entity(createdSnapshot).build();
```
This endpoint has been refactored and no longer returns a response here.

mtien-apache commented on nifi-registry-core/nifi-registry-web-api/src/main/java/org/apache/nifi/registry/web/api/BucketFlowResource.java:
```diff
@@ -291,6 +303,129 @@ public Response createFlowVersion(
+            versionedFlowSnapshot.getSnapshotMetadata().setFlowIdentifier(null);
+            versionedFlowSnapshot.getSnapshotMetadata().setLink(null);
+            versionedFlowSnapshot.getSnapshotMetadata().setVersion(-1);
+            versionedFlowSnapshot.getSnapshotMetadata().setVersion(-1);
```
Good idea - I've changed to use a static variable and removed the duplicate line of code.

mtien-apache commented on nifi-registry-core/nifi-registry-web-api/src/main/java/org/apache/nifi/registry/web/api/BucketFlowResource.java:
```diff
@@ -291,6 +303,129 @@ public Response createFlowVersion(
+        // set remaining snapshot metadata
+        final String userIdentity = NiFiUserUtils.getNiFiUserIdentity();
+        versionedFlowSnapshot.getSnapshotMetadata().setAuthor(userIdentity);
+        versionedFlowSnapshot.getSnapshotMetadata().setVersion(-1);
```
Changed to use a static variable.

mtien-apache commented on nifi-registry-core/nifi-registry-web-api/src/main/java/org/apache/nifi/registry/web/api/BucketFlowResource.java:
```diff
@@ -385,6 +520,64 @@ public Response getLatestFlowVersionMetadata(
+
+        String attachmentName = "flow-version-" + versionNumber;
+
+        UriUtils.encodePath(attachmentName, StandardCharsets.UTF_8);
```
I've removed the encoding.

mtien-apache commented on nifi-registry-core/nifi-registry-web-ui/src/main/webapp/services/nf-registry.api.js:
```diff
@@ -75,6 +75,113 @@ NfRegistryApi.prototype = {
+    exportDropletVersionedSnapshot: function (dropletUri, versionNumber) {
+        var self = this;
+        var url = '../nifi-registry-api/' + dropletUri;
+        url += '/versions/' + versionNumber + '/export';
```
I don't think it's necessary to split up. I have changed to one line.

mtien-apache commented on nifi-registry-core/nifi-registry-web-api/src/main/java/org/apache/nifi/registry/web/api/BucketFlowResource.java:
```diff
@@ -291,6 +303,129 @@ public Response createFlowVersion(
+    @ApiOperation(
+            value = "Create flow",
+            notes = "Creates a flow in the given bucket. The flow id is created by the server and populated in the returned entity.",
+            response = VersionedFlow.class,
```
Changed to VersionedFlowSnapshot.

mtien-apache commented on nifi-registry-core/nifi-registry-web-api/src/main/java/org/apache/nifi/registry/web/api/BucketFlowResource.java:
```diff
@@ -291,6 +303,129 @@ public Response createFlowVersion(
+            @ApiResponse(code = 403, message = HttpStatusMessages.MESSAGE_403),
+            @ApiResponse(code = 404, message = HttpStatusMessages.MESSAGE_404),
+            @ApiResponse(code = 409, message = HttpStatusMessages.MESSAGE_409)})
+    public Response importFlow(
```
My thought for creating this endpoint was to mainly handle the uploaded file and clear the snapshot metadata. It also made it a bit easier for the frontend to only make one request to the server. I have refactored some logic here to minimize duplicating code. It now returns `createFlowVersion` endpoint.

mtien-apache replied:
@bbende I went ahead and removed the `importFlow` method. Now when importing a new flow, a call will first be made to `createFlow`, then another to `importVersionedFlow`.

mtien-apache commented on nifi-registry-core/nifi-registry-web-ui/src/main/webapp/components/explorer/grid-list/dialogs/import-new-flow/nf-registry-import-new-flow.js:
```diff
@@ -0,0 +1,182 @@
+        self.writableBuckets = self.filterWritableBuckets(self.buckets);
+    },
+
+    //TODO: get from the service instead
```
My goal is to have this be like a utility method in Angular service (`nf-registry.service.js`). It would only show the buckets the user is authorized to write. I've been having issues injecting it into this import-new-flow component, so I'm in the middle of troubleshooting that.

mtien-apache commented:
I have pushed a commit that resolves the importing and exporting issues found by @bbende, as well as addressing other feedback found above this comment. 

Thanks for the first reviews, @bbende and @exceptionfactory !

andrewmlim commented:
This new functionality is great! It is going to be extremely useful.

Here are some thoughts after some UI/UX testing:

- I think it would be an improvement if the "Delete data flow" action simply said "Delete flow".  We refer to that type of asset in the Registry as "flow" and the documentation also refers to it as a "flow".   The dialog that appears after selecting the action also says "Delete flow" so this change would help with consistency.
- The title for the "Delete flow" dialog should be changed to "Delete Flow" for case consistency with the "Import New Version" and "Download Version" dialogs.
- When a flow version is downloaded the default name of the file is always flow-version-#.json (where # is the version #). I think it would be helpful if the name would default to the flow name.  So if I had a flow named "Kafka2HDFS" and wanted to download version 3 of it, the default name would be "Kafka2HDFS-version-3.json".

andrewmlim commented:
Additional thoughts:

- In the "Import New Flow" and "Import New Version" dialogs, change "Data Flow Name" to "Flow Name" for same reasons stated earlier.
- In the "Import New Version" dialog, consider changing "Change Log Comment" to "Version Comments" which is what is used in NiFi when saving a flow version.

![version_comments](https://user-images.githubusercontent.com/18287533/115901652-777c8700-a42f-11eb-8a4b-afad517dcd65.png)

andrewmlim commented:
Testing how the feature does error handling and noticed the following:

When importing new flow, if I select a JSON file that is valid JSON, but not a valid flow definition file, after selecting import I get the error message "Cannot create versioned flow snapshot - flowContents must not be null".  After closing the dialog, it seems like nothing was imported, but if you refresh the view a new flow is imported with version 0.  (As an aside, if you select "Download version" on this corrupt flow, the UI seems to open a dialog on the left side of the window.)

![bad_json](https://user-images.githubusercontent.com/18287533/115910486-e14e5e00-a43a-11eb-8ded-03463681aeff.png)

mtien-apache commented:
@andrewmlim Thanks for reviewing! 

- The UI changes make sense to me, especially if the names will match what's in the documentation and NiFi. 
@moranr Do you have any additional thoughts before I proceed?

- For the "DELETE" flow button color in the 'Delete Flow' dialog is what's currently in Registry:
> * I noticed that the color of the "DELETE" button in the "Delete Flow" dialog is a different color (more red) than the other dialog buttons (IMPORT and DOWNLOAD). This appears to be intentional. The "DELETE" button for deleting a bucket is the same red, so understandable if the color difference is to make it stand out since it is a unrecoverable action.

- I'm looking into the invalid flow definition file error now.

moranr commented:
Hey @mtien-apache - thanks for all your work on this. @andrewmlim 's comments all make sense to me and I would proceed with his recommendations. RE the Delete button color, yes the vibrant red is intentional as Drew observed. I believe this color is also used in other "delete" scenarios, so perhaps double-check to make sure color usage is consistent.

mtien-apache commented:
> Hey @mtien-apache - thanks for all your work on this. @andrewmlim 's comments all make sense to me and I would proceed with his recommendations. RE the Delete button color, yes the vibrant red is intentional as Drew observed. I believe this color is also used in other "delete" scenarios, so perhaps double-check to make sure color usage is consistent.

@moranr Thanks for confirming - will proceed with the UI changes. @andrewmlim Confirmed the Delete button color is the same as other Delete buttons used throughout Registry (eg, Delete Buckets).

mtien-apache commented:
Pushed 2 commits: 

- Handled the scenario when an invalid snapshot file is uploaded during Import New Flow @andrewmlim 
- Refactored client side API import methods
     - In the upload file methods, I removed `formData` and now passing the file as an API param. Also using HTTP headers to pass additional data
     - In `uploadFlow`, I'm now making multiple calls to the server that re-uses the existing `BucketFlowResource.createFlow` method
- Refactored server side import methods
     - Removed `BucketFlowResource.importFlow`
     - Refactored logic from `BucketFlowResource.importVersionedFlow` to the service facade
   
UI changes: 
- Fixed remaining UI style issues
- In the 'Import New Flow' dialog, the initial value for the Bucket dropdown menu is set to the Bucket the user is currently viewing. Otherwise, no value is set when viewing all buckets.
- Added messages to inform the user why the 'Import New Flow' button is displayed
     - One displays a message in the center of the page when there are no existing buckets 
     - Another message is displayed under the button buckets exist but the user does not have any write permissions

andrewmlim commented:
Testing the latest changes and noticed that whenever I select "Import new version", the version is always shown as "v.2" in the "Import New Version" dialog even if is it the 3rd, 4th, 5th, etc. version that I am importing.

andrewmlim commented:
In the Actions drop-down, "Import new version" is always highlighted even when another selection has the focus.

![Actions_highlight](https://user-images.githubusercontent.com/18287533/116733672-60eba800-a9ba-11eb-847d-a6d5486d1c7f.png)

andrewmlim commented:
> Related to my last comment, here are the messages in the error dialogs I have found so far:
> 
> * Non flow definition JSON file: "Cannot create versioned flow snapshot - flowContents must not be null"
> * Incomplete flow definition JSON file: "Deserialization of uploaded JSON failed"
> 
> It would be great if we could make these error messages more user friendly. But that could definitely be done as a separate Jira/PR. I wouldn't block this PR to address this.

For the incomplete flow definition JSON file, the error I see now when importing is: "Unexpected end-of-input: expected close marker for Object (start marker at [Source: (org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$UnCloseableInputStream); line: 1, column: 1]) at [Source: (org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$UnCloseableInputStream); line: 2, column: 1]"

which is more specific to the issue (that there is no closing } in the JSON), but less readable.

mtien-apache commented:
> > Related to my last comment, here are the messages in the error dialogs I have found so far:
> > 
> > * Non flow definition JSON file: "Cannot create versioned flow snapshot - flowContents must not be null"
> > * Incomplete flow definition JSON file: "Deserialization of uploaded JSON failed"
> > 
> > It would be great if we could make these error messages more user friendly. But that could definitely be done as a separate Jira/PR. I wouldn't block this PR to address this.

@andrewmlim I refactored the resource method to let the Spring framework to automatically handle the deserialization and removed the "Deserialization of uploaded JSON failed" message. 

Agree to improve the error message in a separate PR for the "Cannot create versioned flow snapshot - flowContents must not be null" message.

> For the incomplete flow definition JSON file, the error I see now when importing is: "Unexpected end-of-input: expected close marker for Object (start marker at [Source: (org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$UnCloseableInputStream); line: 1, column: 1]) at [Source: (org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$UnCloseableInputStream); line: 2, column: 1]"
> 
> which is more specific to the issue (that there is no closing } in the JSON), but less readable.

After investigating this exception message, it seems the exception is being caught by an `ExceptionMapper` and translating it to a BadRequest. It may be from `jackson-databind` in the Spring framework, which is not something we have much control over. It's possible to write a custom ExceptionMapper for a BadRequest, but the custom message could end up applying across the app for any method throwing that type of Exception, which can be addressed in a separate PR.

Although I agree the error message could be better, it at least points to which line the error is found. I believe this is sufficient for the time being. Thoughts?

andrewmlim commented:
> 
> Although I agree the error message could be better, it at least points to which line the error is found. I believe this is sufficient for the time being. Thoughts?

I agree that it is sufficient for now. All of the error messages can be handled outside of this PR. Thanks!

mtien-apache commented:
Pushed a commit:
- Added a new `ExportedVersionedFlowSnapshot` object
- Fixed Import New Version dialog to show the next version to be imported
- Refactored frontend and server side code.
- Added frontend and backend integration tests.

Will add more frontend tests.

mtien-apache commented:
> In the Actions drop-down, "Import new version" is always highlighted even when another selection has the focus.
> 
> ![Actions_highlight](https://user-images.githubusercontent.com/18287533/116733672-60eba800-a9ba-11eb-847d-a6d5486d1c7f.png)

@andrewmlim (cc: @moranr ) It seems like the initial focus behavior is based on keyboard actions and from an accessibility perspective, it's the expected behavior. When using the arrow keys, the focus moves with the new selection. I think what is making it confusing is that the keyboard and mouseover selections have the same red highlighting styles. I believe this highlighting behavior was always present but probably wasn't an issue since there had only been 1 menu item (Delete). 

One proposed fix is I can dig in to change the styles between a keyboard selection and mouseover selection, something like in other dropdown menus in Registry:

<img width="400" alt="Screen Shot 2021-05-04 at 11 29 44 AM" src="https://user-images.githubusercontent.com/56892372/117064443-111f2080-acdb-11eb-86c9-0c26df6031bf.png">

Thoughts?

exceptionfactory commented on nifi-registry-core/nifi-registry-web-api/src/main/java/org/apache/nifi/registry/web/api/BucketFlowResource.java:
```diff
@@ -291,6 +294,42 @@ public Response createFlowVersion(
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    @ApiOperation(
+            value = "Upload flow version",
```
Should this be changed to Import flow version?
```suggestion
            value = "Import flow version",
```

exceptionfactory commented on nifi-registry-core/nifi-registry-web-api/src/main/java/org/apache/nifi/registry/web/api/BucketFlowResource.java:
```diff
@@ -291,6 +294,42 @@ public Response createFlowVersion(
+    @Produces(MediaType.APPLICATION_JSON)
+    @ApiOperation(
+            value = "Upload flow version",
+            notes = "Uploads the next version of a flow. The version number of the object being created must be the " +
+                    "next available version integer. Flow versions are immutable after they are created.",
```
The notes make it sound like the version number must be specified, as opposed to being generated.
```suggestion
            notes = "Import the next version of a flow. The version number of the object being created will be the " +
                    "next available version integer. Flow versions are immutable after they are created.",
```

exceptionfactory commented on nifi-registry-core/nifi-registry-web-api/src/main/java/org/apache/nifi/registry/web/service/StandardServiceFacade.java:
```diff
@@ -360,6 +362,48 @@ public VersionedFlowSnapshot getLatestFlowSnapshot(final String flowIdentifier)
+
+        // if there are new comments, then set it
+        // otherwise, keep the original comments
+        if (!StringUtils.isBlank(comments)) {
```
This could be changed to `StringUtils.isNotBlank()`
```suggestion
        if (StringUtils.isNotBlank(comments)) {
```

exceptionfactory commented on nifi-registry-core/nifi-registry-web-api/src/main/java/org/apache/nifi/registry/web/service/StandardServiceFacade.java:
```diff
@@ -360,6 +362,48 @@ public VersionedFlowSnapshot getLatestFlowSnapshot(final String flowIdentifier)
+        // otherwise, keep the original comments
+        if (!StringUtils.isBlank(comments)) {
+            metadata.setComments(comments);
+        } else if (versionedFlowSnapshot.getSnapshotMetadata() != null && versionedFlowSnapshot.getSnapshotMetadata().getComments() != null) {
```
Should this line check for blank comments?
```suggestion
        } else if (versionedFlowSnapshot.getSnapshotMetadata() != null && StringUtils.isNotBlank(versionedFlowSnapshot.getSnapshotMetadata().getComments())) {
```

exceptionfactory commented on nifi-registry-core/nifi-registry-web-api/src/test/java/org/apache/nifi/registry/web/api/FlowsIT.java:
```diff
@@ -541,4 +544,171 @@ public void testFlowNameUniquePerBucket() throws Exception {
+        flowSnapshot.setSnapshotMetadata(flowSnapshotMetadata);
+        flowSnapshot.setFlowContents(new VersionedProcessGroup()); // an empty root process group
+        flowSnapshot.getFlowContents().setName("Test Flow name");
+        flowSnapshot.getSnapshotMetadata().setVersion(-1);
```
Recommend setting a private static final variable named LATEST_VERSION and reusing here and in the other test method.

exceptionfactory commented on nifi-registry-core/nifi-registry-web-ui/src/main/webapp/components/explorer/grid-list/dialogs/import-new-flow/nf-registry-import-new-flow.js:
```diff
@@ -0,0 +1,192 @@
+
+        // if there's only 1 writable bucket, always set as the initial value in the bucket dropdown
+        // if opening the dialog from the explorer/grid-list, there is no active bucket
+        if (typeof this.activeBucket === 'undefined') {
```
Can this be simplified to check against `undefined` without `typeof`?
```suggestion
        if (this.activeBucket === undefined) {
```

exceptionfactory commented on nifi-registry-core/nifi-registry-web-ui/src/main/webapp/services/nf-registry.api.js:
```diff
@@ -75,6 +75,137 @@ NfRegistryApi.prototype = {
+        return this.http.get(url, options).pipe(
+            map(function (response) {
+                // export the VersionedFlowSnapshot by creating a hidden anchor element
+                var stringSnapshot = encodeURIComponent(JSON.stringify(response.body, null, 2));
```
Is it necessary to call `JSON.stringify()` or is it possible to just get the text content of response.body?  If `response.body` is already JSON, that means it went through an unnecessary serialization step, which would be great to avoid if possible.

mtien-apache replied:
@exceptionfactory Good catch! I can return a `responseType: 'text'` and omit calling `JSON.stringify()`. The file content will not be in pretty print format but that seems ok to me.

exceptionfactory commented on nifi-registry-core/nifi-registry-web-api/src/main/java/org/apache/nifi/registry/web/api/BucketFlowResource.java:
```diff
@@ -291,6 +294,42 @@ public Response createFlowVersion(
+            @ApiParam(value = "The flow identifier")
+            final String flowId,
+            @ApiParam("file") final VersionedFlowSnapshot versionedFlowSnapshot,
+            @HeaderParam("comments") final String comments) {
```
Recommend capitalizing this header to follow HTTP conventions:
```suggestion
            @HeaderParam("Comments") final String comments) {
```

exceptionfactory reviewed (commented):
The recent updates look good @mtien-apache!  I noted a few additional minor recommendations.

bbende commented:
Latest changes look good!

One minor comment... when downloading a version, the window that pops up has the latest version labeled as "Latest Version (v.#)" but then the other items in the list are just "Version #", seems a little inconsistent that one has "v." and the others don't. Should the latest one just say "Latest Version (#)" or maybe "Version # (latest)" ?

Also, while writing this, I started thinking about how we have generally been referring to this as import/export, and the buttons for import use the word "Import", but the for export we are calling it "Download Version". I'm not too hung up on this, but started wondering if that should be called "Export Version" to better correspond with "Import"?

andrewmlim commented:
> One proposed fix is I can dig in to change the styles between a keyboard selection and mouseover selection, something like in other dropdown menus in Registry:
> 
> <img alt="Screen Shot 2021-05-04 at 11 29 44 AM" width="400" src="https://user-images.githubusercontent.com/56892372/117064443-111f2080-acdb-11eb-86c9-0c26df6031bf.png">
> 
> Thoughts?

That would be nice if you could change the styles. That would definitely help distinguish the two difference selections.

andrewmlim commented:
I tested a secure Registry with different authorizations given to users (on the users themselves and also on bucket policies). Auth looks good!

andrewmlim commented:
> One minor comment... when downloading a version, the window that pops up has the latest version labeled as "Latest Version (v.#)" but then the other items in the list are just "Version #", seems a little inconsistent that one has "v." and the others don't. Should the latest one just say "Latest Version (#)" or maybe "Version # (latest)" ?
> 
> Also, while writing this, I started thinking about how we have generally been referring to this as import/export, and the buttons for import use the word "Import", but the for export we are calling it "Download Version". I'm not too hung up on this, but started wondering if that should be called "Export Version" to better correspond with "Import"?

I think the most consistent label would be "Latest (Version #)".

"Export Version" is a good suggestion.  Like Bryan, I don't think that change is critical.

moranr commented:
> I think the most consistent label would be "Latest (Version #)".

@mtien-apache  I like that suggestion.

> "Export Version" is a good suggestion. Like Bryan, I don't think that change is critical.

Also agree here, but not critical. Typically you see Upload/Download or Import/Export used in together.

mtien-apache commented on nifi-registry-core/nifi-registry-web-ui/src/main/webapp/components/explorer/grid-list/dialogs/download-versioned-flow/nf-registry-download-versioned-flow.js:
```diff
@@ -0,0 +1,96 @@
+        var version = this.selectedVersion;
+
+        this.nfRegistryApi.exportDropletVersionedSnapshot(this.droplet.link.href, version).subscribe(function (response) {
+            if (!response.status || response.status === 200) {
```
@exceptionfactory A successful response has typically been just the response body returned. So, when the body is returned (a success) there will not be a status property. I can see how this could be confusing at first glance and without much context. This if-statement is used throughout the codebase, so I decided to follow the conventions.

mtien-apache commented:
Pushed commits that addresses all remaining changes and PR feedback. The PR is ready for a full review. 
@sardell Would you review the frontend changes?

Recent changes include:

- Changed the dialog title from 'Download Version' to 'Export Version' and refactored the component. @bbende @andrewmlim @moranr Thanks for the feedback and agreed 'Export' seems more consistent  

- Changed the 'Export Version' dialog dropdown menu item to "Latest (Version #)"
- Changed the Actions menu hover style to differentiate from the focus style
- Added frontend tests. Note - With the way Angular and Javascript is currently configured in Registry, I had issues writing certain tests for the new components. It's been captured in a JIRA ticket: [NIFIREG-454](https://issues.apache.org/jira/browse/NIFIREG-454). Once resolved, more tests can be added.

exceptionfactory commented on nifi-registry-core/nifi-registry-web-api/src/main/java/org/apache/nifi/registry/web/api/BucketFlowResource.java:
```diff
@@ -291,6 +294,42 @@ public Response createFlowVersion(
+            }
+    )
+    @ApiResponses({
+            @ApiResponse(code = 400, message = HttpStatusMessages.MESSAGE_400),
```
Adding an `ApiResponse` annotation with a code of `201` would be helpful for completeness in the generated documentation.

exceptionfactory commented on nifi-registry-core/nifi-registry-web-api/src/main/java/org/apache/nifi/registry/web/api/BucketFlowResource.java:
```diff
@@ -291,6 +294,42 @@ public Response createFlowVersion(
+
+        final VersionedFlowSnapshot createdSnapshot = serviceFacade.importVersionedFlowSnapshot(versionedFlowSnapshot, bucketId, flowId, comments);
+        publish(EventFactory.flowVersionCreated(createdSnapshot));
+        return Response.status(Response.Status.CREATED).entity(createdSnapshot).build();
```
Best practice when returning an HTTP 201 includes adding a `Location` header with the URL of the new resource.  See `ApplicationResource.generateCreatedResponse()` as a helper method for returning the Location header with the created entity.

mtien-apache replied:
@exceptionfactory I've added the header. Thanks!

exceptionfactory reviewed (commented):
Thanks for working through the feedback details @mtien-apache!  This looks just about good to go, I noted one more detail around the HTTP 201 Created response.

andrewmlim commented:
> Recent changes include:
> 
> * Changed the dialog title from 'Download Version' to 'Export Version' and refactored the component. @bbende @andrewmlim @moranr Thanks for the feedback and agreed 'Export' seems more consistent
> * Changed the 'Export Version' dialog dropdown menu item to "Latest (Version #)"
> * Changed the Actions menu hover style to differentiate from the focus style

These look good! Thanks for making all of these improvements @mtien-apache !

sardell commented:
@mtien-apache I just finished looking at the UI code and testing your changes in the UI. Nice work. I'm a +1 (non-binding).

mtien-apache commented:
Closed in favor of https://github.com/apache/nifi/pull/5107
